### PR TITLE
enhanced monitor, process commands

### PR DESCRIPTION
**Pull request by haristku**
### Description:

[breaking down the previous PR to specific part](https://github.com/matomo-org/plugin-QueuedTracking/pull/259)

- enhanced monitor, process commands

### CODE REVIEW

**Review by snake14 — CHANGES_REQUESTED** [review]
The Process changes look good and work as expected :+1: 

When I tried testing the Monitor changes, the paging didn't appear to work. It would insert new lines above the window height and the visible portion of the window didn't change. It appeared to be appending the help text after some of the lines as well.

**Comment by haristku** [comment]
@snake14
what OS are you using to run the Monitor command? do you use *nix terminal or windows PowerShell or CMD?

unfortunately, i'm not running matomo in a Windows environment, so i don't have a chance to see the "Monitor" output in a Windows console like PowerShell or CMD.. 

🙏 ✌️

**Comment by snake14** [comment]
> @snake14 what OS are you using to run the Monitor command? do you use *nix terminal or windows PowerShell or CMD?
> 
> unfortunately, i'm not running matomo in a Windows environment, so i don't have a chance to see the "Monitor" output in a Windows console like PowerShell or CMD..
> 
> 🙏 ✌️

@haristku I'm running Ubuntu 22.04 and using the stock terminal.

**Comment by haristku** [comment]
> The Process changes look good and work as expected 👍
> 
> When I tried testing the Monitor changes, the paging didn't appear to work. It would insert new lines above the window height and the visible portion of the window didn't change. It appeared to be appending the help text after some of the lines as well.

what if you add 3 rows per page args: 

`php console queuedtracking:monitor -p 3` 

what's the result? is the paging working or still not?

___
the default rows per page is 16, if we have 16 workers it means we only have 1 page, so the paging feature won't work, but if we set the workers to 34, it means we have 3 pages and the paging feature will work, and we can use the right, left, up, down arrows to navigate through the page. the total pages is displayed at the bottom `Q [0-15] | page 1/3 | press (0-9.,q) or arrow(L,R,U,D) | diff/sec 0`

and the help description also included in: 

` php console help queuedtracking:monitor`
```text
Description:
  Shows and updates the current state of the queue every 2 seconds.
  Key ,=first page, .=last page, 0-9=move to page section, arrow LEFT=prev page, RIGHT=next page, UP=next 10 pages, DOWN=prev 10 pages, q=quit

-p, --perpage=PERPAGE                Number of queue worker displayed per page. [default: 16]
```

___
we already split the PR into smaller parts, in this PR we cannot have workers higher than 16, therefore we can use `-p` parameter to simulate pages

___
i still want to update this PR to remove the use of PCNTL completely... i'll update it tomorrow...


ty.

**Review by snake14 — APPROVED** [review]
Thank you for the suggestion of changing the page size. That helped me test some variations. Although, I tested without setting a specific page size again and it worked correctly. I'm not sure what the issue was last time I tested. I'm approving the PR :+1:

File: Commands/Process.php [hunk 1]
```diff
@@ -76,29 +79,84 @@ protected function doExecute(): int
...
+                $lastTimeGotMoreThanZeroTrackedReq = microtime(true);
+            }
+
+            if (!$infiniteCycle) {
+                $numberOfProcessCycle--;
+            }
+            if ($numberOfProcessCycle > 0 || $infiniteCycle) {
+                $cTogo = $infiniteCycle ? "infinite" : $numberOfProcessCycle;
+                $output->writeln("===========================================================================");
+                $output->writeln("<comment>Taking a nap for {$napster} second(s) before re-run the process. <info>({$cTogo})</info> cyle(s) to go.</comment>");
```
Comment (AltamashShaikh) [inline Commands/Process.php]:
```suggestion
                $output->writeln("<comment>Taking a nap for {$napster} second(s), before re-running the process. <info>({$cTogo})</info> cyle(s) to go.</comment>");
```

**Review by AltamashShaikh — APPROVED** [review]
Looks okay to me, left  1 comment.

**Comment by haristku** [comment]
> Looks okay to me, left 1 comment.

`pcntl` usage removed. the consequence is blinking cursor on monitor command… coz cursor hide and show mechanism cannot be restored without pcntl sigterm.

@AltamashShaikh @snake14 , have you ever tried the monitor command in a Windows environment? 
i hope the output and functionality can be the same in all environments.

**Comment by snake14** [comment]
> > Looks okay to me, left 1 comment.
> 
> `pcntl` usage removed. the consequence is blinking cursor on monitor command… coz cursor hide and show mechanism cannot be restored without pcntl sigterm.
> 
> @AltamashShaikh @snake14 , have you ever tried the monitor command in a Windows environment? i hope the output and functionality can be the same in all environments.

@haristku Thank you for all of your effort. I think that the PR looks ready to merge. Sadly, I don't think we currently have a Windows environment to test with. @AltamashShaikh Do you think that we're alright to merge anyway?

**Comment by AltamashShaikh** [comment]
@snake14 We can merge this and address the concern for windows if any

**Comment by haristku** [comment]
> @haristku Thank you for all of your effort. I think that the PR looks ready to merge. Sadly, I don't think we currently have a Windows environment to test with. @AltamashShaikh Do you think that we're alright to merge anyway?

thanks for creating matomo and this plugin in the first place. 👍

btw, i forgot to remove these [2 lines](https://github.com/matomo-org/plugin-QueuedTracking/blob/832db07eceac66855ea9c74874bdc105a03f4ee8/Commands/Monitor.php#L40), my intention is to clear the screen, but it seems unnecessary, you can remove them.✌️

**Comment by AltamashShaikh** [comment]
@haristku Did you mean [this](https://github.com/matomo-org/plugin-QueuedTracking/blob/832db07eceac66855ea9c74874bdc105a03f4ee8/Commands/Monitor.php#L40..L41) 2 lines ?

**Comment by haristku** [comment]
> @haristku Did you mean [this](https://github.com/matomo-org/plugin-QueuedTracking/blob/832db07eceac66855ea9c74874bdc105a03f4ee8/Commands/Monitor.php#L40..L41) 2 lines ?

yes...

**Comment by haristku** [comment]
and 1 more update, change this [line](https://github.com/matomo-org/plugin-QueuedTracking/blob/832db07eceac66855ea9c74874bdc105a03f4ee8/Commands/Monitor.php#L107) to this:

```php
$diffSumInQueue = $diffSumInQueue < 0 ? "<fg=red;options=bold>" . number_format(abs($diffRps)) . "</>" : "<fg=green;options=bold>" . number_format($diffRps) . "</>";
```

✌️

**Comment by AltamashShaikh** [comment]
@haristku All done and released v5.1.0 of the plugin.

**Comment by haristku** [comment]
ty sir... 🫡

**Pull request merged**